### DISCUSSION TITLE
V4 discussion

### DISCUSSION THREAD

**Discussion by jcs224** [jcs224/oak_sessions#25] [Ideas]
This is a thread for V4 of oak_sessions. There are some breaking changes I'd like to make, hence the major version increase.

- Avoid storing session state (even temporarily) in session class instance (addressed in PR #23, still evaluating)
- Make CookieStore encryption mandatory
- require `deleteSession` to be executed from within a session context, without the option of passing context or session ID as an argument (simplifies codebase, also in above PR)
- Remove third-party code (Luxon, Crypto JS)

**Comment by jcs224** [comment]
@c0per :point_up:

**Comment by c0per** [comment]
- I think we should use the native [crypto](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/encrypt) web api provided by deno instead of CryptoJS
- Same for the `luxon` in session expiration. Native `Date` is enough for this use case.

Reducing these not so necessary deps will make the code base stronger in my mind. Both of them won't be a breaking change or invalidating existing sessions. We still use aes encryption and date string to ISO standards.

**Reply by jcs224** [reply]
Sounds good. Especially with Deno Deploy having a 20MB code limit, any reductions on third-party libs are helpful.

**Comment by c0per** [comment]
#23 and #27 are the two possible under consideration. They only differs in `Session.initMiddleware()` api and ways to set `SessionOptions`.

After the decision and merging into the main, I think I can make 2 PRs reducing `CryptoJS` and `luxon` deps (using native `crypto` and `Date` instead).

**Reply by jcs224** [reply]
Awesome! I merged in both our PRs (there were no conflicts, since my changes built on top of yours, and you get credit for the first PR :smile:)

I created #28 and #29. I don't think these warrant preventing a v4 release, so if these libraries aren't removed in the coming days, we can remove them in a patch release later.

**Reply by c0per** [reply]
Yes. Thank you.

**Comment by c0per** [comment]
I'm working on #28 and #29. These are just small changes.

**Reply by jcs224** [reply]
Great work so far. 

I have to write some migration docs, then I'll cut a new release, sometime this week probably.

Thanks for your help!

**Comment by jcs224** [comment]
V4 is launched! Thanks @c0per for your great contributions!
